core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -645,14 +660,21 @@ public void close()
             pageBufferRetainedSizeInBytes = 0;
             bufferedPageCount = 0;
             writeBuffer = null;
-            try (Closer closer = Closer.create()) {
-                closer.register(exchange);
-                if (exchangeSink != null) {
-                    closer.register(exchangeSink::abort);
+
+            if (exchangeSink != null && !exchangeSinkFinished.get()) {
+                try {
+                    exchangeSink.abort().whenComplete((result, failure) -> {
```

Comment:
TBH I feel like `java.util.concurrent.Future#cancel` is a better way (thread safe) to handle cancellation. 

Otherwise, why to bother with `exchangeSinkFinished.set(true);`. `!exchangeSinkFinished.get()` is racy with https://github.com/trinodb/trino/pull/10699/commits/8415672457b4ac9dd963749b15b08cae8175b376#diff-b366169ac97dd5dbdbbe1691dc5054236f8e24f451abb8a298d26e4b811d213cR616. I would just remove it.

You probably should check that `finish` hasn't started yet

Reply:
> Otherwise, why to bother with exchangeSinkFinished.set(true); check? !exchangeSinkFinished.get() is racy with 8415672#diff-b366169ac97dd5dbdbbe1691dc5054236f8e24f451abb8a298d26e4b811d213cR616. I would just remove it.

Basically we only want to `abort` exchange sink if it hasn't been finished already. It can happen when query is cancelled. If the `ExchangeSink` is finished there's no reason to abort.

If the `finish` is still in process it is still legit to request `abort`. 

Cancelling the `Finish` future shouldn't be necessary. If the `ExchangeSink` is aborted before `Finish` is done the `ExchangeSink` implementation may decide to never finish the `finish` operation internally.

Reply:
I would just do
```
            if (exchangeSink != null) {
                try {
                    // abort any pending transfers if there are any
                    exchangeSink.abort().whenComplete((result, failure) -> {
```
but I'm fine either way